### PR TITLE
[WIP] Preserving isTrans data

### PR DESCRIPTION
**Pull request by devedse**
### Prerequisites

- [X] I have written a descriptive pull-request title
- [X] I have verified that there are no overlapping [pull-requests](https://github.com/SixLabors/ImageSharp/pulls) open
- [X] I have verified that I am following matches the existing coding patterns and practice as demonstrated in the repository. These follow strict Stylecop rules :cop:.
- [ ] I have provided test coverage for my change (where applicable)

### Description

Currently still WIP but I wanted to contribute something to the project :smile:. So I'm working on fixing my own issue:
#763

### CODE REVIEW

**Comment by devedse** [comment]
I'm not sure if we want to implement it this way or have a look at the already existing Palette code which should be extended to always write alpha if the hasAlpha flag is true. But please let me know :)

**Comment by JimBobSquarePants** [comment]
Hey @devedse this all looks great! Thanks for making the effort to fix this for us. 

I'll pull it down for a final review but from looking on GitHub I can't see anything amiss. 

In the interim could you please sign the CLA? Cheers! 👍

**Comment by devedse** [comment]
I think the bitconversion could use some improvements

File: src/ImageSharp/Formats/Png/PngMetaData.cs [hunk 1]
```diff
@@ -42,6 +44,31 @@ private PngMetaData(PngMetaData other)
         /// </summary>
         public float Gamma { get; set; }
 
+        /// <summary>
+        /// Gets or sets the Rgb 24 transparent color. This represents any color in an 8 bit Rgb24 encoded png that should be transparent
+        /// </summary>
+        public Rgb24? Rgb24Trans { get; set; }
```
Comment (antonfirsov) [inline src/ImageSharp/Formats/Png/PngMetaData.cs]: I think we should not use abbrevations in public API-s. `TransparentRgb24`? @JimBobSquarePants any better idea?

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs [hunk 2]
```diff
@@ -300,6 +305,52 @@ public void Encode<TPixel>(Image<TPixel> image, Stream stream)
...
+        /// <summary>
+        /// Writes the transparency markers to the stream
+        /// </summary>
+        /// <param name="stream">The <see cref="Stream"/> containing image data.</param>
+        /// <param name="pngMetaData">The image meta data.</param>
+        private void WriteTransparencyMarkers(Stream stream, PngMetaData pngMetaData)
+        {
+            if (pngMetaData.ColorType == PngColorType.Rgb)
+            {
+                if (pngMetaData.Rgb48Trans != null)
```
Comment (JimBobSquarePants) [inline src/ImageSharp/Formats/Png/PngEncoderCore.cs]: We should be checking the bit depth here also.

File: src/ImageSharp/MetaData/ImageMetaData.cs [hunk 3]
```diff
@@ -5,6 +5,7 @@
 using SixLabors.ImageSharp.Formats;
 using SixLabors.ImageSharp.MetaData.Profiles.Exif;
 using SixLabors.ImageSharp.MetaData.Profiles.Icc;
+using SixLabors.ImageSharp.PixelFormats;
```
Comment (JimBobSquarePants) [inline src/ImageSharp/MetaData/ImageMetaData.cs]: I don't think we need this.

**Comment by JimBobSquarePants** [comment]
@antonfirsov To answer your questions regarding the tRNS chunk.

When decoding either grayscale,  24bit or 48bit color types, if the chunk is encountered we decode to `Rgba32` and `Rgba64` pixel formats (we don't have a GrayAlpha pixel format) and set the alpha component to fully opaque for any pixel that does not match the value contained within the chunk.

When encoding we use the `Rgb24`, `Rgb48`, `Gray8`, or `Gray16` pixel format layouts and assign a color that should be regarded as transparent to the chunk.

Does that make sense?

File: src/ImageSharp/MetaData/ImageMetaData.cs (hunk 3 above)
Comment (devedse) [reply src/ImageSharp/MetaData/ImageMetaData.cs]: Fixed

Comment (devedse) [reply src/ImageSharp/MetaData/ImageMetaData.cs]: Fixed

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs (hunk 2 above)
Comment (devedse) [reply src/ImageSharp/Formats/Png/PngEncoderCore.cs]: I'm not exactly sure how to do that, might be better if you implement this.

File: src/ImageSharp/Formats/Png/PngMetaData.cs (hunk 1 above)
Comment (devedse) [reply src/ImageSharp/Formats/Png/PngMetaData.cs]: Fixed

File: src/ImageSharp/Formats/Png/PngMetaData.cs [hunk 4]
```diff
@@ -42,6 +44,31 @@ private PngMetaData(PngMetaData other)
...
+
+        /// <summary>
+        /// Gets or sets the 8 bit grayscale transparent color. This represents any color in an 8 bit grayscale encoded png that should be transparent
+        /// </summary>
+        public byte? LuminanceTrans { get; set; }
+
+        /// <summary>
+        /// Gets or sets the 16 bit grayscale transparent color. This represents any color in a 16 bit grayscale encoded png that should be transparent
+        /// </summary>
+        public ushort? Luminance16Trans { get; set; }
```
[the remark this answers is not in the document]
Comment (devedse) [reply src/ImageSharp/Formats/Png/PngMetaData.cs]: Fixed

File: src/ImageSharp/Formats/Png/PngMetaData.cs [hunk 5]
```diff
@@ -42,6 +44,31 @@ private PngMetaData(PngMetaData other)
...
+
+        /// <summary>
+        /// Gets or sets the Rgb 48 transparent color. This represents any color in a 16 bit Rgb24 encoded png that should be transparent
+        /// </summary>
+        public Rgb48? Rgb48Trans { get; set; }
+
+        /// <summary>
+        /// Gets or sets the 8 bit grayscale transparent color. This represents any color in an 8 bit grayscale encoded png that should be transparent
+        /// </summary>
+        public byte? LuminanceTrans { get; set; }
```
[the remark this answers is not in the document]
Comment (devedse) [reply src/ImageSharp/Formats/Png/PngMetaData.cs]: Fixed

**Comment by devedse** [comment]
I've implemented almost all findings, please redo code review :)

**Comment by antonfirsov** [comment]
> and set the alpha component to fully opaque for any pixel that does not match the value contained within the chunk

But what happens when we encounter pixels matching the value? Do we decode fully transparent pixels? Do we alter transparent pixels to chunk-defined pixels when encoding?

**Comment by JimBobSquarePants** [comment]
@antonfirsov 

>But what happens when we encounter pixels matching the value? Do we decode fully transparent pixels in those cases?

We set the alpha component to 0 leaving the other component data intact.

>Do we alter transparent pixels to chunk-defined pixels when encoding?

No the tRNS chunk is simply a marker. Pixels are encoded as normal. Delete the tRNS chunk and you have an opaque image.

@devedse 

We have some test images already that might cover your needs but there's an official suite here.

http://www.schaik.com/pngsuite/

**Comment by devedse** [comment]
> @devedse
> 
> We have some test images already that might cover your needs but there's an official suite here.
> 
> http://www.schaik.com/pngsuite/

What do you mean by this? For writing test cases?

**Comment by antonfirsov** [comment]
@devedse input images for decoder tests.

File: src/ImageSharp/Formats/Png/PngEncoderCore.cs (hunk 2 above)
Comment (JimBobSquarePants) [reply src/ImageSharp/Formats/Png/PngEncoderCore.cs]: Actually I think it's ok as we already throw if the combination is invalid.

**Comment by JimBobSquarePants** [comment]
Tests added and correct behaviour confirmed. I think this is good to go now.

**Pull request merged**